### PR TITLE
feat: add thinking_budget (version 2)

### PR DESCRIPTION
**Pull request by thyecust**
## Motivation

See https://github.com/sgl-project/sglang/pull/6089

## Modifications

Removed incorrect diff in protocol.py. See commit `version 2`.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by thyecust** [comment]
@zhyncs May you provide more information about https://github.com/sgl-project/sglang/pull/6181#issuecomment-2869190561 ? I cannot reproduce that.

**Comment by thyecust** [comment]
```
12 results - 4 files

sglang • python/sglang/srt/model_executor/model_runner.py:
   ...

sglang • python/sglang/srt/openai_api/adapter.py:
   557                  "min_new_tokens": request.min_tokens,
   558:                 "thinking_budget": request.thinking_budget,
   559                  "stop": request.stop,

  1130              "min_new_tokens": request.min_tokens,
  1131:             "thinking_budget": request.thinking_budget,
  1132              "stop": stop,

sglang • python/sglang/srt/sampling/sampling_batch_info.py:
   ...

sglang • python/sglang/srt/sampling/sampling_params.py:
   ...
```

There are only two references of `request.thinking_budget`, one for  `ChatCompletionRequest` and one for `CompletionRequest`. The definitions is in `protocol.py`. So I think they are safe.

**Comment by thyecust** [comment]
@minleminzui   Could you help review this? thx

**Comment by minleminzui** [comment]
> ```
> 12 results - 4 files
> 
> sglang • python/sglang/srt/model_executor/model_runner.py:
>    ...
> 
> sglang • python/sglang/srt/openai_api/adapter.py:
>    557                  "min_new_tokens": request.min_tokens,
>    558:                 "thinking_budget": request.thinking_budget,
>    559                  "stop": request.stop,
> 
>   1130              "min_new_tokens": request.min_tokens,
>   1131:             "thinking_budget": request.thinking_budget,
>   1132              "stop": stop,
> 
> sglang • python/sglang/srt/sampling/sampling_batch_info.py:
>    ...
> 
> sglang • python/sglang/srt/sampling/sampling_params.py:
>    ...
> ```
> 
> There are only two references of `request.thinking_budget`, one for `ChatCompletionRequest` and one for `CompletionRequest`. The definitions is in `protocol.py`. So I think they are safe.

@CatherineSue, @ispobock, @sleepcoo please help to review this pr, please ensure it doesn't affect your internal services.

**Review by CatherineSue — COMMENTED** [review]
It seems we allow user to pass this parameter with a non-thinking model? I think it is better to isolate this from non-thinking model to prevent potential effects in the future

**Comment by thyecust** [comment]
> It seems we allow user to pass this parameter with a non-thinking model? I think it is better to isolate this from non-thinking model to prevent potential effects in the future

in `sampling_batch_info.py`

```python
        if any(hasattr(r.tokenizer, "think_end_id") for r in reqs): 
             think_end_ids = torch.tensor( 
                 [getattr(r.tokenizer, "think_end_id", -1) for r in reqs], 
                 dtype=torch.int64, 
             ).to(device, non_blocking=True) 
             num_thinking_tokens = torch.tensor([0 for _ in reqs], dtype=torch.int64).to( 
                 device, non_blocking=True
```

For a non-thinking model, r.tokenizer has no think_end_id attribute, so nothing will happen.

Better isolation introduces more code. Is this silent ignorance acceptable? @CatherineSue

File: python/sglang/srt/sampling/sampling_batch_info.py [hunk 1]
```diff
@@ -146,6 +169,35 @@ def from_schedule_batch(cls, batch: ScheduleBatch, vocab_size: int):
...
+            self.num_thinking_tokens - 1 > self.thinking_budgets
+        )
+        next_token_logits.masked_fill_(should_stop.unsqueeze(0), float("-inf"))
+        batch_indices = torch.nonzero(should_stop, as_tuple=True)[0]
+        if len(batch_indices) > 0:
+            end_token_indices = self.think_end_ids[batch_indices]
+            next_token_logits[batch_indices, end_token_indices] = 0.0
+
+    def update_thinking_budgets(self, next_token_ids: torch.Tensor):
+        if not torch.any(self.thinking_budgets > 0):
```
Comment (ispobock) [inline python/sglang/srt/sampling/sampling_batch_info.py]: check self.thinking_budgets is None

File: python/sglang/srt/sampling/sampling_batch_info.py [hunk 2]
```diff
@@ -146,6 +169,35 @@ def from_schedule_batch(cls, batch: ScheduleBatch, vocab_size: int):
     def __len__(self):
         return len(self.temperatures)
 
+    def apply_thinking_budgets(self, next_token_logits: torch.Tensor):
+        has_budget = self.thinking_budgets > 0
```
Comment (ispobock) [inline python/sglang/srt/sampling/sampling_batch_info.py]: check self.thinking_budgets is None

File: python/sglang/srt/sampling/sampling_batch_info.py [hunk 3]
```diff
@@ -76,7 +81,22 @@ def from_schedule_batch(cls, batch: ScheduleBatch, vocab_size: int):
         min_ps = torch.tensor(
             [r.sampling_params.min_p for r in reqs], dtype=torch.float
         ).to(device, non_blocking=True)
-
+        if any(hasattr(r.tokenizer, "think_end_id") for r in reqs):
```
Comment (ispobock) [inline python/sglang/srt/sampling/sampling_batch_info.py:84]: For non-reasoning model, we can skip this check? Do we need to identify if the model is a reasoning model from the architect?

File: python/sglang/srt/openai_api/adapter.py [hunk 4]
```diff
@@ -1127,6 +1128,7 @@ def v1_chat_generate_request(
             "temperature": request.temperature,
             "max_new_tokens": request.max_tokens or request.max_completion_tokens,
             "min_new_tokens": request.min_tokens,
+            "thinking_budget": request.thinking_budget,
```
Comment (ispobock) [inline python/sglang/srt/openai_api/adapter.py]: It seems only valid when `enable_thinking = True`. Need to do the validation.

File: python/sglang/srt/openai_api/protocol.py [hunk 5]
```diff
@@ -172,6 +172,7 @@ class CompletionRequest(BaseModel):
     top_k: int = -1
     min_p: float = 0.0
     min_tokens: int = 0
+    thinking_budget: Optional[int] = None
```
Comment (ispobock) [inline python/sglang/srt/openai_api/protocol.py]: Will it be used in non-chat scenario? Is there any usage example/reference for that?

File: python/sglang/srt/sampling/sampling_batch_info.py (hunk 3 above)
Comment (thyecust) [reply python/sglang/srt/sampling/sampling_batch_info.py:84]: I am thinking about it. Now I don't know how to decide whether a model is reasoning from a request.

File: python/sglang/srt/sampling/sampling_batch_info.py (hunk 2 above)
Comment (thyecust) [reply python/sglang/srt/sampling/sampling_batch_info.py]: thanks. done in https://github.com/sgl-project/sglang/pull/6208/commits/6aca64c2fdae45eaf8e3cb640d4d57e58d24eec1

Comment (thyecust) [reply python/sglang/srt/sampling/sampling_batch_info.py:175]: thanks. done in https://github.com/sgl-project/sglang/pull/6208/commits/6aca64c2fdae45eaf8e3cb640d4d57e58d24eec1

File: python/sglang/srt/sampling/sampling_batch_info.py (hunk 1 above)
Comment (thyecust) [reply python/sglang/srt/sampling/sampling_batch_info.py]: thanks. done in https://github.com/sgl-project/sglang/commit/6aca64c2fdae45eaf8e3cb640d4d57e58d24eec1

File: python/sglang/srt/openai_api/protocol.py (hunk 5 above)
Comment (thyecust) [reply python/sglang/srt/openai_api/protocol.py]: I am not familiar with CompletionRequest. Removed `thinking_budget` in it https://github.com/sgl-project/sglang/pull/6208/commits/186b1aae27c09739ce24fad4cc6c5144f58999a5

File: python/sglang/srt/openai_api/adapter.py (hunk 4 above)
Comment (thyecust) [reply python/sglang/srt/openai_api/adapter.py]: done in https://github.com/sgl-project/sglang/pull/6208/commits/1c741a60134144ebc2303de06f0f7bdc6478e1bc

File: python/sglang/srt/sampling/sampling_batch_info.py (hunk 3 above)
Comment (thyecust) [reply python/sglang/srt/sampling/sampling_batch_info.py:84]: As far as I can think of, there seems to be no better way. The information about can be obtained here is very limited. It is just doing sampling and does not know the specific model architecture.

File: python/sglang/srt/sampling/sampling_batch_info.py [hunk 6]
```diff
@@ -146,6 +169,37 @@ def from_schedule_batch(cls, batch: ScheduleBatch, vocab_size: int):
     def __len__(self):
         return len(self.temperatures)
 
+    def apply_thinking_budgets(self, next_token_logits: torch.Tensor):
+        if self.thinking_budgets is None:
+            return
+        has_budget = self.thinking_budgets > 0
```
Comment (lambert0312) [inline python/sglang/srt/sampling/sampling_batch_info.py]: Doesn't support `=0` scenario?

File: python/sglang/srt/sampling/sampling_batch_info.py [hunk 7]
```diff
@@ -146,6 +169,37 @@ def from_schedule_batch(cls, batch: ScheduleBatch, vocab_size: int):
...
+            self.num_thinking_tokens - 1 > self.thinking_budgets
+        )
+        next_token_logits.masked_fill_(should_stop.unsqueeze(0), float("-inf"))
+        batch_indices = torch.nonzero(should_stop, as_tuple=True)[0]
+        if len(batch_indices) > 0:
+            end_token_indices = self.think_end_ids[batch_indices]
+            next_token_logits[batch_indices, end_token_indices] = 0.0
+
+    def update_thinking_budgets(self, next_token_ids: torch.Tensor):
+        if self.thinking_budgets is None or not torch.any(self.thinking_budgets > 0):
```
Comment (lambert0312) [inline python/sglang/srt/sampling/sampling_batch_info.py]: Doesn't support `=0` scenario?

Comment (thyecust) [reply python/sglang/srt/sampling/sampling_batch_info.py]: Hi, now we support `thinking_budget=0` in https://github.com/sgl-project/sglang/pull/6208/commits/30aa15fb455c0cd06f91655edc192d3d675164c3

File: python/sglang/srt/sampling/sampling_batch_info.py (hunk 6 above)
Comment (thyecust) [reply python/sglang/srt/sampling/sampling_batch_info.py]: Hi, now we support `thinking_budget=0` in https://github.com/sgl-project/sglang/commit/30aa15fb455c0cd06f91655edc192d3d675164c3

File: python/sglang/srt/sampling/sampling_batch_info.py [hunk 8]
```diff
@@ -76,7 +81,22 @@ def from_schedule_batch(cls, batch: ScheduleBatch, vocab_size: int):
...
+        if any(hasattr(r.tokenizer, "think_end_id") for r in reqs):
+            think_end_ids = torch.tensor(
+                [getattr(r.tokenizer, "think_end_id", -1) for r in reqs],
+                dtype=torch.int64,
+            ).to(device, non_blocking=True)
+            num_thinking_tokens = torch.tensor([0 for _ in reqs], dtype=torch.int64).to(
+                device, non_blocking=True
+            )
+            thinking_budgets = torch.tensor(
+                [r.sampling_params.thinking_budget or -1 for r in reqs],
```
Comment (lambert0312) [inline python/sglang/srt/sampling/sampling_batch_info.py]: When `thinking_budget=0`, this will be assigned a value of -1. It is recommended to modify this, otherwise it will not be fully supported.

Comment (thyecust) [reply python/sglang/srt/sampling/sampling_batch_info.py]: updated https://github.com/sgl-project/sglang/pull/6208/commits/200c5984aa81e6d862644169e876d145c69f035a
thanks for pointing out

**Comment by hnyls2002** [comment]
Thanks @thyecust! `thinking_budget` is now supported natively on `main`, so I'm closing this. If your use case isn't covered, please feel free to reopen.
